### ISSUE TITLE
[question] Using conan.tools.microsoft.MSBuild

### ISSUE THREAD

**Issue by KingKiller100** [conan-io/conan#12979]
Conan Version: 1.57.0

Hi,

I'm converting my conanfile.py scripts to use the recommended Microsoft generators and MSBuild tools. Migrating the generators has been okay but I have noticed something when migrating the MSBuild over. I noticed that by default, the build doesn't utilize the parallel compilations offered by MSBuild. Looking at your src, I noticed you need a value in your conan.conf or profile settings for the `tools.microsoft.msbuild:max_cpu_count` to allow for the `/m` switch to be appended and enable the parallel build.

I didn't need to do this before so I was wondering if there is a way for the conan environment to detect the number of cpus and automatically fill the value for this?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @KingKiller100 

Thanks for your question.
First, please note that the ``conf`` ``tools.build:jobs`` is already being used to define the ``/MP`` paralellism, and this is the one that is defaulted to use the number of CPUs. This is the reason why ``tools.microsoft.msbuild:max_cpu_count`` is not defaulted, as using the number of CPUs for that easily saturates the current machine. If you are using it, you might want to reduce ``tools.build:jobs``.

Profiles are jinja templates, if you want to parameterize the value of some conf, it is possible. The modules ``platform`` and ``os`` are injected and can be used in the jinja template, you might try to define it from an env-var for example, like ``os.getenv("MYCPUCOUNT")``?

**Comment by KingKiller100** [comment]
Hi,

Currently, I do not have defined the `tools.build:jobs` setting anywhere. Would I expect MSBuild to activate the parallel building without defining this setting?

Would I need to place this setting, `tools.build:jobs`, in each of the profile files or the `conan.conf`?

**Comment by memsharded** [comment]
> Currently, I do not have defined the tools.build:jobs setting anywhere. Would I expect MSBuild to activate the parallel building without defining this setting?

Yes, the default is that if it is not defined, it will already enable parallelism via /MP to the current number of cpus

It can be defined in profiles ``[conf]`` section, in the ``global.conf`` or in command line ``-c ....``

**Comment by KingKiller100** [comment]
Do these settings play an effect on the build/package id of the package? 

Also I'm only familiar with the `conan.conf` file that holds user conan settings. What is this `global.conf`? Another name for the file I mentioned

**Comment by memsharded** [comment]
No, that is the thing: ``conf`` by default do not affect ``package_id`` of the package (as the number of cores to build shouldn't produce a different binary at all).

The ``conan.conf`` file is deprecated (still some bits necessary in 1.X), being superseded by a new file called ``global.conf``, in the same location of the COnan cache (the root of the cache). The ``global.conf`` can store all the new ``conf`` items. Actually they can be defined in:

- ``global.conf`` file (can also be managed with ``conan config install``
- Profiles ``[conf]`` section (can be different for build/host context too)
- Command line ``-c ...`` argument (can be different for build/host context too)

**Comment by memsharded** [comment]
Closing this ticket as answered back then and staled.
Please open new tickets for any further question or issue if necessary, thanks for your feedback!
